File: src/dataset/cpp/phone_number_Exercism_thefullarcticfox/claude/ClaudeSonnet4_phone_number_v1.cpp
#include "phone_number.h"
#include <stdexcept>

namespace phone_number {
	phone_number::phone_number(const std::string& phone_number) {
		_number.reserve(10);
		
		for (char digit : phone_number) {
			if (std::isdigit(digit)) {
				_number.push_back(digit);
			} else if (digit != ' ' && digit != '.' && digit != '+' && 
					   digit != '-' && digit != '(' && digit != ')') {
				throw std::domain_error("invalid number");
			}
		}
		
		const size_t size = _number.size();
		if (size < 10 || size > 11)
			throw std::domain_error("invalid number length");
		
		if (size == 11) {
			if (_number[0] != '1')
				throw std::domain_error("invalid country code");
			_number.erase(0, 1);
		}
		
		if (_number[0] < '2')
			throw std::domain_error("invalid area code");
		if (_number[3] < '2')
			throw std::domain_error("invalid exchange code");
	}

	const std::string&	phone_number::number() const {
		return _number;
	}

	std::string			phone_number::area_code() const {
		return _number.substr(0, 3);
	}

	phone_number::operator std::string() const {
		std::string result;
		result.reserve(14);
		result += '(';
		result.append(_number, 0, 3);
		result += ") ";
		result.append(_number, 3, 3);
		result += '-';
		result.append(_number, 6, 4);
		return result;
	}
}  // namespace phone_number

File: src/dataset/go/phone_number_Exercism_go_rootulp/claude/ClaudeSonnet4_phone_number_v1.go
package phonenumber

import (
	"fmt"
	"strings"
	"unicode"
)

func AreaCode(phoneNumber string) (string, error) {
	number, err := Number(phoneNumber)
	if err != nil {
		return "", err
	}
	return number[0:3], nil
}

func ExchangeCode(phoneNumber string) (string, error) {
	number, err := Number(phoneNumber)
	if err != nil {
		return "", err
	}
	return number[3:6], nil
}

func SubscriberNumber(phoneNumber string) (string, error) {
	number, err := Number(phoneNumber)
	if err != nil {
		return "", err
	}
	return number[6:], nil
}

func Number(phoneNumber string) (string, error) {
	number, err := clean(phoneNumber)
	if err != nil {
		return "", err
	}
	if number[0] < '2' {
		return "", fmt.Errorf("area code %s can not start with 0 or 1", number[0:3])
	}
	if number[3] < '2' {
		return "", fmt.Errorf("exchange code %s can not start with 0 or 1", number[3:6])
	}
	return number, nil
}

func clean(phoneNumber string) (string, error) {
	cleaned := make([]byte, 0, 11)
	for _, character := range phoneNumber {
		if unicode.IsDigit(character) {
			cleaned = append(cleaned, byte(character))
		}
	}
	if len(cleaned) == 11 && cleaned[0] == '1' {
		cleaned = cleaned[1:]
	}
	if len(cleaned) != 10 {
		return "", fmt.Errorf("phone number %s must have 10 digits", string(cleaned))
	}
	return string(cleaned), nil
}

func Format(phoneNumber string) (string, error) {
	number, err := Number(phoneNumber)
	if err != nil {
		return "", err
	}
	var result strings.Builder
	result.Grow(14)
	result.WriteByte('(')
	result.WriteString(number[0:3])
	result.WriteString(") ")
	result.WriteString(number[3:6])
	result.WriteByte('-')
	result.WriteString(number[6:])
	return result.String(), nil
}